src/stores/MessagePreviewStore.ts

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2020 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { Room } from "matrix-js-sdk/src/models/room";
+import { ActionPayload } from "../dispatcher/payloads";
+import { AsyncStoreWithClient } from "./AsyncStoreWithClient";
+import defaultDispatcher from "../dispatcher/dispatcher";
+import { RoomListStoreTempProxy } from "./room-list/RoomListStoreTempProxy";
+import { textForEvent } from "../TextForEvent";
+import { MatrixEvent } from "matrix-js-sdk/src/models/Event";
+import { _t } from "../languageHandler";
+
+const PREVIEWABLE_EVENTS = [
+    // This is the same list from RiotX
+    {type: "m.room.message", isState: false},
+    {type: "m.room.name", isState: true},
+    {type: "m.room.topic", isState: true},
+    {type: "m.room.member", isState: true},
+    {type: "m.room.history_visibility", isState: true},
+    {type: "m.call.invite", isState: false},
+    {type: "m.call.hangup", isState: false},
+    {type: "m.call.answer", isState: false},
+    {type: "m.room.encrypted", isState: false},
+    {type: "m.room.encryption", isState: true},
+    {type: "m.room.third_party_invite", isState: true},
+    {type: "m.sticker", isState: false},
+    {type: "m.room.create", isState: true},
+];
+
+interface IState {
+    [roomId: string]: string;
+}
+
+export class MessagePreviewStore extends AsyncStoreWithClient<IState> {
+    private static internalInstance = new MessagePreviewStore();
+
+    private constructor() {
+        super(defaultDispatcher, {});
+    }
+
+    public static get instance(): MessagePreviewStore {
+        return MessagePreviewStore.internalInstance;
+    }
+
+    /**
+     * Gets the pre-translated preview for a given room
+     * @param room The room to get the preview for.
+     * @returns The preview, or null if none present.
+     */
+    public getPreviewForRoom(room: Room): string {
+        if (!room) return null; // invalid room, just return nothing
+
+        // It's faster to do a lookup this way than it is to use Object.keys().includes()
+        const val = this.state[room.roomId];
+        if (val !== null && typeof(val) !== "string") {
```

Comment:
according to IState iface it is only ever a string so when would it not be a string (other than undefined) and when would it be null?

Reply:
undefined = not present, it's a map. null = always a possibility, and in this case means the message is empty.